### PR TITLE
Object pose and size estimation with superquadric-model

### PR DESCRIPTION
**Pull request by towardthesea**
Hi,
This `PR` is to introduce a new feature to the `IOL` and `OPC` pipeline, which provides better size and pose estimation of an object.

Following is an example
[iol2opc-with-superquadric](https://www.youtube.com/watch?v=dwNk11og5I8)

## Dependencies:
- [superquadric-model](https://github.com/robotology/superquadric-model)

## How to use:
- Turn on the new parameter **use_superquadric** in [iol2opc configuration file](https://github.com/robotology/icub-client/blob/feature/iol2opc_with_superquadric/app/iol2opc/conf/config.ini#L9)
- Run **superquadric-model** and make all connections in this [application](https://github.com/robotology/superquadric-model/blob/master/app/scripts/superquadric-model.xml.template)
- Run all **iol2opc** related modules and make connection.

*Note: if the *superquadric-model* is not available for connections to *iol2opc* modules, it will use the old way.

Otherwise, the pipeline keeps following the old way to estimate the size and pose through SFM only.

P/S: due to the deadline of the submission, it is not necessary to include this feature in the paper. However this feature is an improvement. 

Thanks and Regards,
Phuong

### CODE REVIEW

**Comment by pattacini** [comment]
Well done @towardthesea 

Just a couple of quick comments:
- It seems there's a lag between the geometry extraction that uses the superquadrics (the geometry is recognized correctly) and the update of the GUI
- Did you try to gauge how the new feature behaves when there are multiple objects (say 3, 4 objects in the scene)?

**Comment by pattacini** [comment]
cc @giuliavezzani

File: src/modules/iol2opc/include/module.h [hunk 1]
```diff
@@ -324,6 +329,10 @@ class IOL2OPCBridge : public RFModule, public iol2opc_IDL
      */
     CvPoint getBlobCOG(const Bottle &blobs, const int i);
 
+    bool    getBlobPoints(const CvPoint &cog, deque<CvPoint> &blobPoints);
```
Comment (Tobias-Fischer) [inline src/modules/iol2opc/include/module.h]: Please add short documentation :)

File: src/modules/iol2opc/include/module.h [hunk 2]
```diff
@@ -324,6 +329,10 @@ class IOL2OPCBridge : public RFModule, public iol2opc_IDL
...
+
+    bool    getSuperQuadric(const CvPoint &point, Vector &pos, Vector &dim);
```
Comment (Tobias-Fischer) [inline src/modules/iol2opc/include/module.h]: Please add short documentation :)

**Comment by Tobias-Fischer** [comment]
Ahhh, and one more thing. Did you try whether the new feature works fine with Karma? As far as I understand, it should .. as the `Subsystem_KARMA` uses the `tableheight+offset` for the `z` coordinate rather than the `object_center`, is it?

**Comment by Tobias-Fischer** [comment]
And one more thing: Could you please change the `icub-client/app/demos/proactiveTagging.xml` to include the superquadric modules?

File: src/modules/iol2opc/include/module.h (hunk 1 above)
Comment (towardthesea) [inline src/modules/iol2opc/include/module.h]: Done for this [here](https://github.com/robotology/icub-client/commit/ba2dd3273568f984df62f437a7e4e5d7d9a4a014)

File: src/modules/iol2opc/include/module.h (hunk 2 above)
Comment (towardthesea) [inline src/modules/iol2opc/include/module.h]: This also :smile:

**Comment by Tobias-Fischer** [comment]
Included this in the paper :+1:. Well done!

**Comment by towardthesea** [comment]
> And one more thing: Could you please change the icub-client/app/demos/proactiveTagging.xml to include the superquadric modules?

Check [it](https://github.com/robotology/icub-client/pull/40/commits/45c443741bf60c4fbfe8260b86f6132afe41953c) @Tobias-Fischer :smile:

**Comment by Tobias-Fischer** [comment]
All good from my side. But yes, @pattacini is right, we should check why there is such a big delay between the change within the superquadric-model and when it changes within `iCubGUI` before merging.

**Comment by towardthesea** [comment]
Hi @Tobias-Fischer @pattacini 

> It seems there's a lag between the geometry extraction that uses the superquadrics (the geometry is recognized correctly) and the update of the GUI.

This is because of the [filter for the pose and dimension of iol2opc](https://github.com/robotology/icub-client/blob/feature/iol2opc_with_superquadric/src/modules/iol2opc/src/module.cpp#L929) . If we turn off this filter, the object size obtained (and shown in iCubGui) will fluctuate, due to the noise blob from **lbpExtract**. So the lag is somehow expected.

> Did you try to gauge how the new feature behaves when there are multiple objects (say 3, 4 objects in the scene)?

Without the [internal filter of superquadric-model](https://github.com/robotology/superquadric-model/blob/master/app/conf/config.ini#L16) we can deal with multiple objects. However please consider the fact that if objects are too closed to each other, the blobs from **lbpExtract** tend to merge and we will get bad result from **superquadric-model**

Cheers,
Phuong

**Comment by pattacini** [comment]
@towardthesea thanks for your further investigation!

The lag is expected as soon as we enable some sort of filtering. If filtering is necessary, then there's no problem in having the lag.

**Pull request merged**